### PR TITLE
docs: fullstack dspy

### PR DESCRIPTION
**Pull request by mikeldking**
Adds a fullstack DSPy example app that lets you do zero_shot and compiled few-shot prompting using DSPy.
Screenshot 2024-04-25 at 4 01 53 PM

### CODE REVIEW

**Comment by axiomofjoy** [comment]
Can nix the cell for OpenAI API keys.

**Review by axiomofjoy — APPROVED** [review]

File: python/examples/dspy-rag-fastapi/frontend/about.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    menu_items=None,
+)
+
+st.title("RAG powered by DSPy")
+
+st.sidebar.info("Select a demo above.")
+
+st.markdown(
+    """
+    ## Description This project introduces a [Streamlit](https://streamlit.io)
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/frontend/about.py:17]:
```suggestion
    ## Description
    
    This project introduces a [Streamlit](https://streamlit.io)
```

File: python/examples/dspy-rag-fastapi/README.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+# Full-Stack DSPy Application with FastAPI and Streamlit
+
+## Introduction
+
+This project is a full-stack application designed to leverage natural language processing capabilities entirely locally and to integrate with the [DSPy](https://github.com/stanfordnlp/dspy) framework developed by StanfordNLP. It features a [FastAPI](https://github.com/tiangolo/fastapi) backend for processing and a [Streamlit](https://streamlit.io) frontend for interactive user interfaces. This implementation is fully local, integrating cutting-edge technologies such as OpenAI for language and embedding models, [Chroma DB](https://github.com/chroma-core/chroma) for vector storage, and [Arize Phoenix](https://github.com/Arize-ai/phoenix) for observability. This setup ensures all operations, from processing to data storage, are executed on the local machine, enhancing privacy, data security, and ease of use.
+
+## Features
+
+-   **Fully Local Execution**: Ensures privacy and security by running all processes on your local machine without external dependencies.
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/README.md:9]: Not sure what this refers to?

File: python/examples/dspy-rag-fastapi/frontend/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Use an official Python runtime as a parent image
+FROM python:3.10-slim
+
+# Install gcc
+RUN apt-get update -y \
+    && apt-get install -y python3-dev \
+    gcc \
+    libc-dev \
+    libffi-dev
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/frontend/Dockerfile:9]:
```suggestion
FROM python:3.10-bullseye
```

This probably works.

File: python/examples/dspy-rag-fastapi/backend/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Use an official Python runtime as a parent image
+FROM python:3.10-slim
+
+# Install gcc
+RUN apt-get update -y \
+    && apt-get install -y python3-dev \
+    gcc \
+    libc-dev \
+    libffi-dev
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/backend/Dockerfile:9]:
```suggestion
FROM python:3.10-bullseye
```

This probably works.

File: python/examples/dspy-rag-fastapi/README.md (hunk 2 above)
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/README.md:9]: will remove

**Pull request merged**